### DISCUSSION TITLE
How do we use different FacialRecognition models e.g., Evision.Face.LBPHFaceRecognizer

### DISCUSSION THREAD

**Discussion by faizann** [cocoa-xu/evision#262] [Q&A] [answered]
I am trying to replicate [this blog post](https://medium.com/@siucy814/training-a-facial-recognition-model-by-opencv-e4717d86b7ec) in Python which uses `LBPHFaceRecognizer`. 
In Python code once you create an instance of `LBPHFaceRecognizer` you get to call train function on it.
There is no such function in Elixir bindings in LBPHFaceRecognizer. Does this mean that it has to be manually coded or have I missed something?

**Comment by cocoa-xu** [comment]
Hi @faizann, the module you're looking for in Elixir is `[Evision.Face.LBPHFaceRecognizer](https://hexdocs.pm/evision/Evision.Face.LBPHFaceRecognizer.html)`. If you're using smart cell in Livebook, then it's not there yet. However, you can follow the similar approach as the `Evision.FaceRecognizerSF`

**Reply by faizann** [reply]
Hi @cocoa-xu  

Thanks for your quick response. I am not using SmartCell rather just elixir app or livebook. 
I have looked at `Evision.FaceRecognizerSF` where features are taken and then can be matched. 

The `Evision.Face.LBPHFaceRecognizer` does not have feature or match or train methods. 

This is from my livebook 
```elixir
face_recogniser = Evision.Face.EigenFaceRecognizer.create()
Evision.Face.FaceRecognizer.train(face_recogniser, [mat],0)
```
Output
```
{:error,
 "cannot get `Ptr<cv::face::FaceRecognizer>` from `self`: mismatched type or invalid resource?"}
```

I feel like I am missing something obvious but couldn't figure it out after spending a few hours on the problem.

**Reply by cocoa-xu** [reply]
Thanks for the error message! I probably know how to address this particular error.

```elixir
{:error,
 "cannot get `Ptr<cv::face::FaceRecognizer>` from `self`: mismatched type or invalid resource?"}
```

I'll try to fix it today or tomorrow and release a new version

**Reply by cocoa-xu** [reply, marked as answer]
Hi @faizann, evision v0.2.8 is shipped and the following code should work :)

```elixir
images = 
  [
    "person0/0.jpg", "person0/1.jpg", "person0/2.jpg",
    "person1/0.jpg", "person1/1.jpg", "person1/2.jpg"
  ]
  |> Enum.map(&Evision.imread(&1, flags: Evision.Constant.cv_IMREAD_GRAYSCALE()))
  |> Enum.map(&Evision.Mat.as_type(&1, :s32))

labels = Nx.tensor([0, 0, 0, 1, 1, 1], type: :s32)

face_recogniser = Evision.Face.EigenFaceRecognizer.create()
Evision.Face.EigenFaceRecognizer.train(face_recogniser, images, labels)
```

**Reply by faizann** [reply]
That is awesome. Thanks a lot for fixing it.
